examples/redis_repl/store/db.go

```diff
@@ -17,26 +17,57 @@ import (
 type redisStore struct {
 	cli      *redis.Client
 	statsCli stats.Client
+	metaDB   uint
 }
 
 func (this *redisStore) Close() error {
-	this.statsCli.Close()
+	_ = this.statsCli.Close()
 	return this.cli.Close()
 }
 
 func isRedisError(err error) bool {
 	return err != nil && !strings.HasSuffix(strings.TrimSpace(err.Error()), "nil")
 }
 
+const (
+	RaftStateKey = "raft.state"
+	RaftStateTermKey = "term"
+	RaftStateIndexKey = "index"
+)
+
+func (this *redisStore) GetLastAppliedEntry() (db.RaftEntry, error) {
+	metaCli := selectDB(int(this.metaDB), this.cli)
+	if result, err := metaCli.HGetAll(RaftStateKey).Result(); err != nil {
+		return db.RaftEntry{}, err
+	} else {
+		term, _ := strconv.ParseUint(result[RaftStateTermKey], 10, 64)
+		index, _ := strconv.ParseUint(result[RaftStateIndexKey], 10, 64)
+		return db.RaftEntry{Term: term, Index: index}, nil
+	}
+}
+
 func (this *redisStore) Load(data []byte) ([]byte, error) {
 	defer this.statsCli.Timing("redis_load.latency.ms", time.Now())
 	luaScript := string(data)
 	return this.evalLua(luaScript)
 }
 
-func (this *redisStore) Save(data []byte) ([]byte, error) {
+const LUAScript =
+`
+redis.call('select', '%d')
+redis.call('hset', '%s', '%s', '%d')
+redis.call('hset', '%s', '%s', '%d')
+%s
+`
+
+func (this *redisStore) Save(raftState db.RaftEntry, data []byte) ([]byte, error) {
 	defer this.statsCli.Timing("redis_save.latency.ms", time.Now())
-	luaScript := string(data)
+	luaSnippet := string(data)
+	luaScript := fmt.Sprintf(LUAScript,
+		this.metaDB,										/* switch to metadata DB */
+		RaftStateKey, RaftStateTermKey, raftState.Term,		/* insert RAFT state term */
+		RaftStateKey, RaftStateIndexKey, raftState.Index,	/* insert RAFT state index */
+		luaSnippet)											/* user supplied Lua snippet */
 	return this.evalLua(luaScript)
```

Comment:
do we not need to switch back to `user db` before executing this statement?

Reply:
At the moment there is an assumption that the user provides the Lua script with the appropriate DB selection followed by the actual keyspace mutations. I'm working on a change that validates this before executing the `luaSnippet`. Once that is in, we don't need to perform any switches. Moreover, its hard to determine which DB to switch to since we don't currently understand what a "user's db" is.

Reply:
The recent changes allow for the user to specify the `db.index` using which we prefix the `luaSnippet` with an appropriate `select` statement. Please check.